test/functional/delegated_keys/test_jce.py

```diff
@@ -49,3 +55,28 @@ def test_generate_correct_key_length(algorithm, requested_bits, expected_bits, l
     test = JceNameLocalDelegatedKey.generate(algorithm, requested_bits)
 
     assert length_finder(test.key) == expected_bits
+
+
+def build_short_key_cases():
+    for algorithm in JAVA_AUTHENTICATOR:
+        if algorithm.upper().startswith("HMAC"):
+            message = "HMAC keys smaller than {} bits are unsafe".format(MinimumKeySizes.HMAC.value)
+            yield (algorithm, MinimumKeySizes.HMAC.value, False, message)
+            yield (algorithm, MinimumKeySizes.HMAC.value - 1, True, message)
+
+        elif algorithm.upper().endswith("RSA"):
+            message = "RSA keys smaller than {} bits are unsafe".format(MinimumKeySizes.RSA.value)
+            yield (algorithm, MinimumKeySizes.RSA.value, False, message)
+            yield (algorithm, MinimumKeySizes.RSA.value // 2, True, message)
+
+    message = "RSA keys smaller than {} bits are unsafe".format(MinimumKeySizes.RSA.value)
+    yield ("RSA", MinimumKeySizes.RSA.value, False, message)
+    yield ("RSA", MinimumKeySizes.RSA.value // 2, True, message)
+
+
+@pytest.mark.parametrize("algorithm, key_bits, too_short, error_message", build_short_key_cases())
+def test_warn_on_short_keys(capturing_logger, algorithm, key_bits, too_short, error_message):
+    _test = JceNameLocalDelegatedKey.generate(algorithm, key_bits)
+
+    logging_results = capturing_logger.getvalue()
+    assert (too_short and error_message in logging_results) or (not too_short and error_message not in logging_results)
```

Comment:
I'm not going to block on this, but nested list comprehensions can make this code hard to parse.

Reply:
@karlw00t That's not a list comprehension, it's just two boolean checks

It just compressed this:
```python
if too_short:
    assert error_message in logging_results
else:
    assert error_message not in logging_results
```
to this:
```python
assert (too_short and error_message in logging_results) or (not too_short and error_message not in logging_results)
```